### ISSUE TITLE
Peer dependencies

### ISSUE THREAD

**Issue by drjonnicholson** [what3words/w3w-node-wrapper#45]
Reopening a new issue as @c5haw closed #44 - the issue is not resolved as CrossFetch is not an _optional_ peer.

Example of the error we get on Node v12 is:

![image](https://user-images.githubusercontent.com/7688837/156021439-1159bf7f-865c-4b71-b409-8b50dad7a95d.png)

We are using Axios, and the expectation is that we should be able to build without also adding the cross fetch dependency. Because the peer dependencies does not specify optionality then to get our application to build we have to install both.

This can be solved by adding meta properties to the package json, which is the approach taken in many projects, such as Storybook: https://github.com/storybookjs/storybook/blob/next/addons/docs/package.json

**Comment by drjonnicholson** [comment]
Sorry, I can't provide that as its a private company repository. But it's essentially a CRACO project. 

From the docs: https://docs.npmjs.com/cli/v8/configuring-npm/package-json#peerdependenciesmeta

The reason some of our colleagues are experiencing this is likely that as they are constrained to Node v12 they probably have NPM <= v6, where according to https://docs.npmjs.com/cli/v8/configuring-npm/package-json#peerdependencies from NPM v7 peer dependencies are automatically installed

**Comment by c5haw** [comment]
I have tried to recreate the issue you're experiencing following the below steps:

1. `npx create-react-app my-app`
2. `cd my-app`
3. `npm add craco @what3words/api cross-fetch`
4. Add `craco.config.js`
```
# craco.config.js
module.exports = {
    reactScriptsVersion: "react-scripts" /* (default value) */,
    webpack: {
        configure: {
          resolve: {
            fallback: {
              // This is only required for webpack 5
              os: false,
            }
          }
        },
    },
};
```
6. Update `App.tsx`
```
# App.tsx
import React, { useEffect, useState } from 'react';
import what3words from '@what3words/api';
import './App.css';

function App() {
  const service = what3words('TSTSTSTS', {}, { transport: 'fetch' })
  const [input, setInput] = useState<string>('');
  useEffect(() => {
    if (input.length > 0) {
      service.autosuggest({ input })
        .then(console.log);
    }
  }, [input]);
  return (
    <div className="App">
      <header className="App-header">
        <input type="text" value={input} onChange={e => setInput(e.currentTarget.value)} />
      </header>
    </div>
  );
}

export default App;
```
7. Update `package.json`
```
# package.json
{
  "name": "my-app",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@testing-library/jest-dom": "^5.16.2",
    "@testing-library/react": "^12.1.3",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.4.1",
    "@types/node": "^16.11.26",
    "@types/react": "^17.0.39",
    "@types/react-dom": "^17.0.11",
    "@what3words/api": "^4.0.5",
    "craco": "^0.0.3",
    "cross-fetch": "^3.1.5",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-scripts": "5.0.0",
    "typescript": "^4.6.2",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "craco start",
    "build": "craco build",
    "test": "craco test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```
8. Ran `npm run build` (using `craco build`) - No errors found. Created a `/build` directory
9. Ran `npm run start` (using `craco start`) - No errors App ran successfully

If you can provide me with a way to recreate the issue you're experiencing then I would be happy to look into it, but unfortunately with the information you've provided thus far I haven't been able to do so using a CRACO project.

**Comment by drjonnicholson** [comment]
Thanks @c5haw, as pointed at in my last post the issue is related to NPM version rather than Node version.

We are seeing this issue with NPM version 6.13.4.

Which version of npm were you using?

**Comment by drjonnicholson** [comment]
Taken from our build server if we remove `cross-fetch` as a dependency (axios is still a dependency):
![image](https://user-images.githubusercontent.com/7688837/159054542-0647c9fd-a5b3-4741-a3b9-15a67aa84251.png)

We are unable to change the build server configuration at this time.

**Comment by drjonnicholson** [comment]
Also note that I tried removing `cross-fetch` locally (clearing and reinstalling dependencies). Nothing in my application mentions `cross-fetch`, but it still gets installed with npm 8.1.0. If I do `npm ls cross-fetch` I can see the peer dependency is automatically getting installed because of the peer dependency defined in this package:
![image](https://user-images.githubusercontent.com/7688837/159055148-5ab16162-151d-4236-a8bf-38496d836089.png)

I suspect if you do the same in your example applications above you'll find that both `cross-fetch` and `axois` are getting installed despite you adding a dependency for one of them.

**Comment by drjonnicholson** [comment]
Hi @c5haw, any update on your end with this?

**Comment by c5haw** [comment]
A fix has been published for this now `@what3words/api@4.0.6` which should resolve the issue you've been experiencing.

**Comment by c5haw** [comment]
Closing this issue as it has grown stale with no activity.
